{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CS 344 Project Final Report\n",
    "## Elizabeth Koning\n",
    "### May 12, 2019\n",
    "\n",
    "\n",
    "## Vision\n",
    "\n",
    "The project is an assessment of using machine learning techniques on the optimization of resistance surfaces.\n",
    "\n",
    "In biology, using computational approaches for the problems of resistance surfaces is relatively new, so while there has been important work done in this area, there is still much space for improvement and assessment.\n",
    "\n",
    "This project addresses the issue from the computational perspective, looking at what has been already implemented to approach the problem and what may be effective to implement in the future.\n",
    "\n",
    "The specific focus of the biological question focuses on deer populations and movement between different populations, and the computational focus is mainly around Genetic Algorithms.\n",
    "\n",
    "\n",
    "## Background\n",
    "\n",
    "### Application Domain\n",
    "\n",
    "The goal of optimizing resistance surfaces is to find the major barriers between populations.\n",
    "\n",
    "Professor Miller's research focuses on deer populations and the barriers that separate them. When physical barriers (such as highways or rivers) are present in an environment, they may cause division into two separate populations. The division prevents mixing and interbreeding. This leads to genetic differences between the populations, which can be examined through sampling. There is the geographic distance between two populations, which can be easily measured. However, the effective distance is measured through the genetic differences. The goal is to use attributes of the landscape in the area that the deer live to reflect the effective distances. Different aspects of the landscape, such as traffic or forest density, may increase the effective distance from the geographic distance by adding resistance.\n",
    "\n",
    "In the past, before introducing computational methods, biologists would look at maps of deer populations and see what the major differences were the aligned with the major differences between populations. For example, they might notice that a river or road ran across the landscape and would logically divide the populations. They may also look at attributes like elevation or locations of cities.\n",
    "\n",
    "Until recently when computational approaches were introduced, the system for identifying the physical barriers has been based on seeing which barriers coorelate with the differences in the population and suggesting that that barrier may be the dividing line. Professor Miller is examining the barriers more statistically and systematically.\n",
    "\n",
    "My work includes evaluating the effectiveness of genetic algorithms for this application and comparing it to the potential behavior of other AI algorithms, including Simulated Annealing and Convolutional Neural Networks.\n",
    "\n",
    "### Relevant Theory\n",
    "\n",
    "Professor Miller's work has been using genetic algorithms to approach this problem, so I will be looking more into how the genetic algorithms work and how they can best be used for this case.\n",
    "\n",
    "#### Genetic Algorithms\n",
    "\n",
    "Genetic algorithms are a machine learning technique that imitate the process of natural selection. This is one of the reasons it is so popular with biologists. The terminology matches that of evolution.\n",
    "\n",
    "The most similar AI approach that we disucussed in class is simulated annealing. Both GA and SA use the selection of random changes for each iteration, requiring improvements increasingly often through the training process.\n",
    "\n",
    "The major pieces of GA are initialization, selection, genetic operators, heuristics, and termination.\n",
    "\n",
    "In intialization, a population hundreds or thousands of possible solutions is created. These are typically randomized.\n",
    "\n",
    "In selection, a part of the population is selected to \"breed\" and create a new generation. A fitness function measures the fitness of each solution so that it is more likely for fit solutions to be selected. The fitness function can be dificult to define, so it may also be a simulation.\n",
    "\n",
    "In the genetic operators step, the selected solutions form the new generations using crossover and mutation. As in biology, two parent solutions combine to form a child solution. Crossover works by combining aspects of each of the parents into a new solution, and mutation works by making random changes to a potential solution. Mutation is important because it keeps diversity within the population. Other genetic operators are possible as well, but crossover and mutation are the most popular.\n",
    "\n",
    "The heuristics can affect the genetic operations processes in order to make the algorithm more efficient.\n",
    "\n",
    "The process of forming more generations is repeated until it reaches a termination condition. This can include reaching a maximum number of generations or plateauing as assessed by the fitness function.\n",
    "\n",
    "(Background information from Wikipedia article on Genetic Algorithms)\n",
    "\n",
    "### Reasoning for Theoretical Foundation Selection\n",
    "\n",
    "#### Existing Work\n",
    "Professor Miller and other biology researchers have been working with Genetic Algorithms on problems addressing resistance surfaces. Peterman from Ohio State, has published ResistanceGA which is an R package specficially designed for this purpose (Peterman 2018).\n",
    "\n",
    "Biologists appreciate genetic algorithms because of their connections with the language and concepts of biology. Genetic algorithms are conceptually easy for them to understand because they are already familiar with both the principles and terminology.\n",
    "\n",
    "R tends to be the programming language that biologists are most familiar with because of its role in statistics.\n",
    "\n",
    "However, genetic algorithms are known to be very slow. If we can achieve similar results with a faster algorithm, that would be helpful. With Miller's deer dataset, it takes 3-5 days to run the genetic algorithm. Even with a small simulated dataset, it takes multiple hours to run.\n",
    "\n",
    "\n",
    "#### Existing Alternative Approaches\n",
    "\n",
    "As this is a problem new to the computational approaches, there is not a wide variety of implementations.\n",
    "\n",
    "There has been research using Gradient Forests, building off of Random Forests, and Generalized dissimilarity modelling (Fitzpatrick 2014). However, their approach is somewhat different, so it would take some work to apply their solutions to these datasets with the resistance surfaces. Specifically, their focus is on entire communities of organisms or between SNPs in a single organism.\n",
    "\n",
    "Beyond this, I read other papers that focus on resistance surfaces or effective distances between populations. However, if they are answering similar questions and using computational methods, then they will be using ResistanceGA.\n",
    "\n",
    "\n",
    "## Implementation\n",
    "\n",
    "In this project, I have run ResistanceGA using simulated data and examined the inputs and outputs of the data.\n",
    "\n",
    "I decided to use simulated data because of time constraints. With the acutal data, running it will take 3-5 days. With the smaller, simulated dataset, it only takes a couple of hours to run.\n",
    "\n",
    "In order to create the simulated dataset, the package creates a \"truth\" map and then samples from there. Then, it it uses the genetic algorithms with the sampled data. As shown in the results, the truth and optimized result graphs match.\n",
    "\n",
    "The Jupyter notebook with the code and results is located at GA/GA.ipynb.\n",
    "\n",
    "After encountering some errors with the non-resilient ResistanceGA package, the code effectively creates the simulated data and runs a genetic algorithm with it. However, it takes multiple hours to run the algorithm, even on such a small dataset.\n",
    "\n",
    "One of the next steps for this project would be to use the same data with alternative algorithms, so the data_for_compare.rmd file has a walkthrough of the data formats as used at each of the different stages and how to export them to use outside of RStudio.\n",
    "\n",
    "\n",
    "## Results\n",
    "\n",
    "Simulated Annealing is likely a great fit because of its similarities with genetic algorithms but better speed. It has many shared properties in terms of the system it uses, but key differences that will affect both its effectiveness and its speed.\n",
    "\n",
    "I would propose using a multiple-start version of simulated annealing. This approach maintains the helpful aspects of GA while also speeding up the process and removing the need for the genetic operations step.\n",
    "\n",
    "A helpful aspect of GA is the wide range of considered solutions. Because of the complexity of the problem and the many landscape attributes that affect the resistance of the surface, it will be helpful to start with a wide population. In the GA, it is the population, but in SA it will be the many starts.\n",
    "\n",
    "The way that the resistance values work with the effective distances makes sense for simulated annelaing and its incremental changes. In this problem, we are looking at how the different landscape attributes are weighed in comparison to one another. It would be reasonable to expect that if the true weight of the influence of traffic is high, but in the current solution it is low, that if the weight is increased a third of the way to the true value, the overall evaluation of that solution will increase instead of decrease.\n",
    "\n",
    "It is even possible that SA could achieve better results than GA. I would not expect there to be many different peaks, so directly climbing up to local maxima could be highly effective in finding the global maxima.\n",
    "\n",
    "While hill climbing is even faster than simulated annealing, I would expect to find local maxima in this sort of problem. In real rather than simulated data, surface attributes such as traffic volume and city categorization will be highly correlated. The way that this would be reflected in the weights of the different landscape attributes would lead to local maxima, making simulated annealing a better option. However, it could also be useful to try an implementation of hill climbing for a more thorough assessment.\n",
    "\n",
    "Another aspect to speed it up will be to use parellization. It is still possible that the GA will still be more effective than the alternatives. In this case, a combination of patience and increased computing power will offer more value than an alternative algorithm. In the case that SA is as or more effective than GA, it is an embarassingly parallel problem.\n",
    "\n",
    "Convolutional Neural Networks seem like they may be a good fit because the data is a set of images, which seems to connect well with CNNs. However, while they intially sounded promising because of the similarites with data as images, there are critically differences. As we discussed in class, CNNs work very well for image recognition and categorization. In this problem with deer populations, we are looking for weights on different aspects of the environment. The way the data is considered is also significantly different. In an image recognition problem, there are many images and each image is a datapoint. In this problem, the dataset is not necessarily smaller, but there are fewer images. Instead, each piece of the image is a datapoint giving information about the cooresponding landscape. Because of these fundamental differences, Neural Networks cannot necessarily be considered ineffective, but CNNs can be ruled out.\n",
    "\n",
    "\n",
    "## Conclusions\n",
    "\n",
    "My recommendation is to further explore the possibilities in simulated annealing. Based on the lack of alternatives to ResistanceGA, there seems to be room for a project seeking to address similar questions but with a more efficient computational approach.\n",
    "\n",
    "I believe the ResistanceGA package could provide value to a project using an alternative algorithm.  A first step I would take would be to consider whether the portion of ResistanceGA that uses the genetic algorithm could be adapted to use simulated annealing instead. The simulated annealing algorithm would still need to create a population of potential solutions, so parts of the code should remain the same, especially the aspects that are specific to the problem.\n",
    "\n",
    "One major loss in moving from genetic algorithms to simulated annealing would be the nature of the analogy. Therefore, another key part of the future work would be articulating simulated annealing in a way to effectively communicate with biologists. Because of the similarities with Genetic Algorithms, some of the language could be similar so long as it didn't imply the two were equivalent, but a key part of the project would need to be finding a metaphor that fit with biology instead of metallurgy. In formulating the new analogy, microbiology or migration may be particularly effective.\n",
    "\n",
    "\n",
    "## Implications\n",
    "\n",
    "The use of machine learning for the optimization of resistance surfaces has implications for the field of biology because it can lead to stronger results through measurement and comparison of the effects of different landscape attributes.\n",
    "\n",
    "Human contributions to the environment are some of the factors in consideration. Knowing how human modifications to the landscape are changing the way different populations of animals relate is key for issues regarding conservation and spread of disease. The newer computational approach to the issue of resistance surfaces means that the effects of different aspects of the landscapes can now be statistically compared."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bibliography\n",
    "\n",
    "Basu, S., Kumbier, K., Brown, J. B., & Yu, B. (2018). Iterative random forests to discover predictiveand stable high-order interactions. Proceedings of the National Academy of Sciences of the United States of America, 115(8), 1943-1948. doi:10.1073/pnas.1711236115\n",
    "\n",
    "Fitzpatrick, M. C., & Keller, S. R. (2014). Ecological genomics meets community-level modelling of biodiversity: mapping the genomic landscape of current and future environmental adaptation. Ecology Letters. doi:10.1111/ele.12376\n",
    "\n",
    "Genetic algorithm. (n.d.). In Wikipedia.\n",
    "\n",
    "Liaw, A., & Wiener, M. (2002, December). Classification and Regression by randomForest. R News, 2(3), 18-22.\n",
    "\n",
    "Peterman, W. E. (2018). ResistanceGA: An R package for the optimization of resistance surfaces using genetic algorithms. Methods Ecol Evol, 9, 1638– 1647. doi:10.1111/2041-210X.12984\n",
    "\n",
    "Ruiz-Lopez, M. J., Barelli, C., Rovero, F., Hodges, K., Roos, C., Peterman, W. E., & Ting, N. (2015). A novel landscape genetic approach demonstrates the effects of human disturbance on the Udzungwa red colobus monkey (Procolobus gordonorum). Heredity, 116(2), 167–176. doi:10.1038/hdy.2015.82\n",
    "\n",
    "Spear, S. F., Balkenhol, N., Fortin, M., McRae, B. H., & Scribner, K. (2010). Use of resistance surfaces for landscape genetic studies: considerations for parameterization and analysis. Molecular Ecology, 19(17), 3576-3591. doi:10.1111/j.1365-294X.2010.04657.x\n",
    "\n",
    "Wade, Alisa A.; McKelvey, Kevin S.; Schwartz, Michael K. 2015. Resistance-surface-based wildlife conservation connectivity modeling: Summary of efforts in the United States and guide for practitioners. Gen. Tech. Rep. RMRS-GTR-333. Fort Collins, CO: U.S. Department of Agriculture, Forest Service, Rocky Mountain Research Station. 93 p."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
